### DISCUSSION TITLE
Inline CSS

### DISCUSSION THREAD

**Discussion by pierredewilde** [withastro/roadmap#343] [Proposal]
Coming back from last year [RFC: a way to inline css #918](https://github.com/withastro/astro/issues/918), I would like to argue the importance of inline CSS, mainly for 2 reasons:

1. inline CSS is faster than external CSS files loaded after HTML page
2. inline CSS 0kb JS require only one request per page while pages with external CSS require several requests per page. Partially hydrated components also increase the number of requests per page. Inline CSS may significantly reduce the number of requests which is cost-effective with Edge deployments where cost is based on it.

Here is a small script used to implement inline CSS:
https://gist.github.com/pierredewilde/9b8b715d4143513c0a41986ea48f9097

**Comment by jasikpark** [comment]
does https://github.com/astro-community/astro-critters not perform that for you?

**Comment by pierredewilde** [comment]
Thank you. It seems to be the case although not yet tested.

**Reply by jasikpark** [reply]
not yet tested? wdym by that exactly?

**Comment by pierredewilde** [comment]
I mean that I have not tested critters yet. It doesn't mean that critters is not tested ;-)

**Reply by jasikpark** [reply]
ah, okay! I can say that I've tested it. I didn't find inlining to be performant enough to warrant in my usage though 👍 

it succeeded at inlining for me though

**Reply by lilnasy** [reply]
In my opinion, using critters should be the last resort for any project. After your page has fully rendered, it parses the just-stringified stylesheet for selectors and then parses the just-stringified markup to match every element against every selector, to see which ones it should keep. It's an O(m * n) time complexity, should be nowhere near SSR.

**Comment by lilnasy** [comment]
I accomplished this by adding a Set to the Astro global. Every compilation of an Astro component adds the styles to the set, and I consume it in the layout component. It gets serialized and put into a Fragment using the `set:html` directive (style tags get removed, so I figured I could use Fragments).

And I think this could be a good pattern: expose the generated stylesheet in the Astro global so the user can decide where to put it.

**Reply by aFuzzyBear** [reply]
Hi @lilnasy  do you have a demo that you could care to share?

**Reply by lilnasy** [reply]
I could put one together if it might help. I am consuming Astro as a library with deno, I just wanted an HTML templating language so idk how much of it applies, but this is what I'm doing:
- astro hands off all transformed styles to Vite, I am injecting them back into the compiled typescript as an export right after the transformation (I said I created a Set earlier, but that didn't play well with caching)
- when a request hits, I am importing the styles along with the matched astro module, and adding it to the `SSRResult.styles` Set.

Unlike my previous way, this way does not expose the styles on `AstroGlobal` and insertion is handled by the server runtime.

This worked neatly for my use case where I took complete control of compilation/routing/rendering, but maybe too vertical of an approach for Astro. In my opinion, the "cheapest" way to incorporate inline styles in the codebase might be by introducing an alternative, less powerful mode for styles that bypasses Vite.

**Reply by aFuzzyBear** [reply]
Oh wow that is a very cool and very edge use case, again it would cooll to see how you are using Astro in this setup, but to address the other part, of bypassing vite for styles we do have the `<style is:raw>` escape hatch which does this for us presently,

**Reply by lilnasy** [reply]
> Oh wow that is a very cool and very edge use case, again it would cooll to see how you are using Astro in this setup, but to address the other part, of bypassing vite for styles we do have the `<style is:raw>` escape hatch which does this for us presently, 

Is it like adding `is:inline` to scripts, in that astro would not do anything to the tag body, not even to scope the selectors.

**Reply by lilnasy** [reply]
@aFuzzyBear I uploaded bit of an MVP to [lilnasy/miniastro](https://github.com/lilnasy/miniastro)

Here's the gist, given this directory structure...

<details>
<summary>src/</summary>

<details>
<summary>pages/</summary>

<details>
<summary>x.astro</summary>

```astro
---
import Layout from '../components/Layout.astro'
import A from '../components/A.astro'
import B from '../components/B.astro'
---
<style>
  h1 {
    font-size: 1rem;
  }
</style>

<Layout title="Home Page">
  <h1>Hello!</h1>
</Layout>
```

</details>

</details>

<details>
<summary>components/</summary>

<details>
<summary>A.astro</summary>

```astro
<style>
  div {
    font-size: 1rem;
  }
</style>
```

</details>

<details>
<summary>B.astro</summary>

```astro
---
import A from './A.astro'
---
<style>
  div {
    font-size: 2rem;
  }
</style>
```

</details>

<details>
<summary>C.astro</summary>

```astro
<style>
  div {
    font-size: 3rem;
  }
</style>
```

</details>

<details>
<summary>Layout.astro</summary>

```astro
---
const { title } = Astro.props
---
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>{title}</title>
  </head>
  <body>
    <slot />
  </body>
</html>
```

</details>

</details>

</details>

...runnning this command will start a server that inlines only the styles needed...

<details>
<summary><code>deno run -A https://github.com/lilnasy/miniastro/raw/d298b7c/main.ts run src/pages</code></summary>

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <title>Home Page</title>
        <style>
            div:where(.astro-YX3ACUVW) {
                font-size: 1rem
            }

            div:where(.astro-P47YP7CL) {
                font-size: 2rem
            }

            h1:where(.astro-OXJX4J4E) {
                font-size: 1rem
            }
        </style>
    </head>
    <body>
        <h1 class="astro-OXJX4J4E">Hello!</h1>
    </body>
</html>
```

</details>

...and running this command will write the generated files to `.miniastro` folder

<details>
<summary><code>deno run -A https://github.com/lilnasy/miniastro/raw/d298b7c/main.ts build src/pages</code></summary>

<details>
<summary>.miniastro/</summary>

<details>
<summary>pages/</summary>

<details>
<summary>server.ts</summary>

```ts
import { serve } from 'https://deno.land/std@0.170.0/http/server.ts'
import { createRouter } from 'https://raw.githubusercontent.com/lilnasy/miniastro/d298b7c94102867c53292fa47f1a5a02673fb203/runtime.ts'
import X, { scripts as X_scripts, styles as X_styles } from './x.astro.ts'

const routes = [
	{ name: 'X', component: X, styles: X_styles, scripts: X_scripts }
]

const router = createRouter(routes)

serve(router)
```

</details>

<details>
<summary>x.astro.ts</summary>

```ts
import {
  Fragment,
  render as $$render,
  createAstro as $$createAstro,
  createComponent as $$createComponent,
  renderComponent as $$renderComponent,
  renderHead as $$renderHead,
  maybeRenderHead as $$maybeRenderHead,
  unescapeHTML as $$unescapeHTML,
  renderSlot as $$renderSlot,
  mergeSlots as $$mergeSlots,
  addAttribute as $$addAttribute,
  spreadAttributes as $$spreadAttributes,
  defineStyleVars as $$defineStyleVars,
  defineScriptVars as $$defineScriptVars,
  
} from "https://esm.sh/astro@1.6.14/runtime/server?target=es2022";
import Layout, { styles as Layout_styles, scripts as Layout_scripts } from '../components/Layout.astro.ts'
import A, { styles as A_styles, scripts as A_scripts } from '../components/A.astro.ts'
import B, { styles as B_styles, scripts as B_scripts } from '../components/B.astro.ts'

const $$Astro = $$createAstro("<stdin>", '', 'file://cas/sdasdaca');
const Astro = $$Astro;

//@ts-ignore
const $$stdin = $$createComponent(async ($$result, $$props, $$slots) => {
const Astro = $$result.createAstro($$Astro, $$props, $$slots);
Astro.self = $$stdin;

return $$render`

${$$renderComponent($$result,'Layout',Layout,{"title":"Home Page","class":"astro-OXJX4J4E"},{"default": () => $$render`${$$maybeRenderHead($$result)}<h1 class="astro-OXJX4J4E">Hello!</h1>`,})}
`;
}, '<stdin>');
export default $$stdin;

export const styles  = [ ...Layout_styles, ...A_styles, ...B_styles, `h1:where(.astro-OXJX4J4E){font-size:1rem}`]
export const scripts = [ ...Layout_scripts, ...A_scripts, ...B_scripts, ]
```

</details>

</details>

<details>
<summary>components/</summary>

<details>
<summary>A.astro.ts</summary>

```ts
import {
  Fragment,
  render as $$render,
  createAstro as $$createAstro,
  createComponent as $$createComponent,
  renderComponent as $$renderComponent,
  renderHead as $$renderHead,
  maybeRenderHead as $$maybeRenderHead,
  unescapeHTML as $$unescapeHTML,
  renderSlot as $$renderSlot,
  mergeSlots as $$mergeSlots,
  addAttribute as $$addAttribute,
  spreadAttributes as $$spreadAttributes,
  defineStyleVars as $$defineStyleVars,
  defineScriptVars as $$defineScriptVars,
  
} from "https://esm.sh/astro@1.6.14/runtime/server?target=es2022";

const $$Astro = $$createAstro("<stdin>", '', 'file://cas/sdasdaca');
const Astro = $$Astro;

//@ts-ignore
const $$stdin = $$createComponent(async ($$result, $$props, $$slots) => {
const Astro = $$result.createAstro($$Astro, $$props, $$slots);
Astro.self = $$stdin;

return $$render`
`;
}, '<stdin>');
export default $$stdin;

export const styles  = [ `div:where(.astro-YX3ACUVW){font-size:1rem}`]
export const scripts = [ ]
```

</details>
<details>

<summary>B.astro.ts</summary>

```ts
import {
  Fragment,
  render as $$render,
  createAstro as $$createAstro,
  createComponent as $$createComponent,
  renderComponent as $$renderComponent,
  renderHead as $$renderHead,
  maybeRenderHead as $$maybeRenderHead,
  unescapeHTML as $$unescapeHTML,
  renderSlot as $$renderSlot,
  mergeSlots as $$mergeSlots,
  addAttribute as $$addAttribute,
  spreadAttributes as $$spreadAttributes,
  defineStyleVars as $$defineStyleVars,
  defineScriptVars as $$defineScriptVars,
  
} from "https://esm.sh/astro@1.6.14/runtime/server?target=es2022";
import A, { styles as A_styles, scripts as A_scripts } from './A.astro.ts'

const $$Astro = $$createAstro("<stdin>", '', 'file://cas/sdasdaca');
const Astro = $$Astro;

//@ts-ignore
const $$stdin = $$createComponent(async ($$result, $$props, $$slots) => {
const Astro = $$result.createAstro($$Astro, $$props, $$slots);
Astro.self = $$stdin;

return $$render`
`;
}, '<stdin>');
export default $$stdin;

export const styles  = [ ...A_styles, `div:where(.astro-P47YP7CL){font-size:2rem}`]
export const scripts = [ ...A_scripts, ]
```

</details>

<details>
<summary>Layout.astro.ts</summary>

```ts
import {
  Fragment,
  render as $$render,
  createAstro as $$createAstro,
  createComponent as $$createComponent,
  renderComponent as $$renderComponent,
  renderHead as $$renderHead,
  maybeRenderHead as $$maybeRenderHead,
  unescapeHTML as $$unescapeHTML,
  renderSlot as $$renderSlot,
  mergeSlots as $$mergeSlots,
  addAttribute as $$addAttribute,
  spreadAttributes as $$spreadAttributes,
  defineStyleVars as $$defineStyleVars,
  defineScriptVars as $$defineScriptVars,
  
} from "https://esm.sh/astro@1.6.14/runtime/server?target=es2022";


const $$Astro = $$createAstro("<stdin>", '', 'file://cas/sdasdaca');
const Astro = $$Astro;

//@ts-ignore
const $$stdin = $$createComponent(async ($$result, $$props, $$slots) => {
const Astro = $$result.createAstro($$Astro, $$props, $$slots);
Astro.self = $$stdin;
const { title } = Astro.props
return $$render`<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>${title}</title>
  ${$$renderHead($$result)}</head>
  <body>
    ${$$renderSlot($$result,$$slots["default"])}
  </body></html>`;
}, '<stdin>');
export default $$stdin;

export const styles  = [ ]
export const scripts = [ ]
```

</details>

</details>

</details>

</details>

**Comment by thdoan** [comment]
My needs were simple, and this setup has worked well for me. YMMV, but this gave me complete control over my inline styles in the simplest way without requiring other dependencies.

**components/InlineStyle.astro**
```
---
---

<slot />
```

**layouts/Layout.astro**
```
---

export interface Props {
  title: string;
}

const { title } = Astro.props;

---

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1, maximum-scale=5">
    <title>{title}</title>
    <link rel="icon" href="/favicon.svg" type="image/svg+xml">
    <slot name="head" />
  </head>
  <body>
    <slot />
  </body>
</html>
```

**pages/index.astro**
```
---

import Layout from '../layouts/Layout.astro';
import InlineStyle from '../components/InlineStyle.astro';

---

<Layout title="Sample">

  <InlineStyle slot="head">
    <style is:inline>
    ...
    </style>
  </InlineStyle>

  <main>
  ...
  </main>

</Layout>
```

If you wanted to minify the inline styles, then you can use [astro-compress](https://www.npmjs.com/package/astro-compress):

**astro.config.mjs**
```
import { defineConfig } from 'astro/config';
import compress from 'astro-compress';

export default defineConfig({
  integrations: [
    compress({
      exclude: [
        // Files in here are already compressed
        '_astro/*'
      ],
      html: {
        // Override html-minifier-terser defaults here
      },
      img: false,
      svg: false,
    })
  ],
});
```

**Reply by lilnasy** [reply]
That's clever and I thought of something similar but it makes you give up astro's scoping and all other processing. 

YSK I opened a PR implementing inline styles natively [#6659](https://github.com/withastro/astro/pull/6659).

**Comment by thdoan** [comment]
@lilnasy I saw your PR and hope it gets merged. Yeah, it sucks that I have to manually prefix my CSS now because autoprefixer doesn't work on `is:inline` :(.

**Comment by lilnasy** [comment]
- Implementation PR: withastro/astro#6659

# Summary

Provide a `build.inlineStylsheets` option to specify inlining behavior of styles authored or imported in astro modules.

# Example

```js
// astro.config.mjs
import { defineConfig } from 'astro'

export default defineConfig({
  build: {
    inlineStylesheets: "always" // or "never", or "auto"
  }
}
```

# Background & Motivation

There has been a constant interest in inlining styles while still taking advantage of scoping and other processing steps since before 1.0 (see: withastro/astro#918), with many users incorrectly assuming that `is:inline` directive is the solution (see: withastro/astro#6388).

Simple one-page websites do not benefit from an external stylesheet, since there is no other page that could use the cached stylesheet. On the other hand, large websites are overoptimized for cacheability, since our chunking splits the styles too granularly. Case in point, Astro docs homepage has 20 linked stylesheets, 14 of them are less than 1kb (see: withastro/astro#6528).

So far we have not provided a way to allow inlining stylesheets, prompting workarounds. However, coming from other frameworks, users might expect to be able to configure this behavior.
- SvelteKit allows configuring a threshold under which CSS files will be inlined:  https://kit.svelte.dev/docs/configuration#inlinestylethreshold
- Fresh lets plugins inject inline styles: https://fresh.deno.dev/docs/concepts/plugins#hook-render
- Nuxt has a global configuration: https://nuxt.com/docs/api/configuration/nuxt-config#inlinessrstyles

# Goals

- Provide a way to reduce the number of HTTP requests for stylesheets.
- Maintain current behavior when not explicitly configured.
- Works with both `server` and `static` outputs.

# Non-Goals

- Identify "critical" CSS rules.
- Enable a single external stylesheet.
- Preloading.

# Detailed Design

The decision to inline a stylesheet should be made at build time (`astro build`), based on the `build.inlineStylsheets` and `vite.build.assetsInlineLimit` configuration options. We already use `assetsInlineLimit` elsewhere to decide if a script should be inlined. `build.inlineStylsheets` option could be set to either `"always"`, `"never"`, or `"auto"`. To stay consistent with the current behavior, the default value should be `"never"`. When `build.inlineStylsheets` is set to `"auto"`, `vite.build.assetsInlineLimit` should be respected. If `assetsInlineLimit` is undefined, a default threshold of 4kb should be used.

If the stylesheet is to be inlined, its contents should be added to `PageBuildData` of each page that it is used in. The asset should also be removed from vite's bundle to avoid unnecessary files in the output.

Astro has run into CSS ordering issues and currently ensures that the `<link>` tags are in a specific order. Currently, this order can't be guaranteed when some stylesheets are to be inlined as `<style>` tags: the rendering pipeline receives the two separtely and inserts all the `<link>` tags first. Therefore, an implementation must pass both inline and external stylesheets in the same array or set that preserves the order, and the rendering pipeline should make sure each `SSRElement` gets serialized with the appropriate tag.

# Testing Strategy

An implementation can be tested adequately with fixture-based tests, ensuring the DOM from SSR responses and generated `.html` files has the expected count of style and link tags and that they are placed in the expected order.

# Drawbacks

- Might lead to unexpected results with our current CSS chunking strategy - it tends to create a lot of tiny stylesheets, and each of them might be under the limit, leading to more CSS being inlined than the user might expect.


# Alternatives

- `astro-critters` integration: it's viable only for static builds of small sites, as it slows to a crawl trying to match every CSS rule against every DOM element on every page.
- `<style critical> ... </style>`: implementation cost, mental overhead for developers, lack of evidence that the granularity will be worth it.

# Adoption Strategy

For now, inlining should be opt-in. In the future, we can consider making "auto" the default.

# Unresolved Questions

- Should the configuration be `build.inlineStylsheets: "never" | "always" | "auto"`? An alternative I've considered is `build.stylesheets: "external" | "inline" | "auto"`.
- Should the user need to set two configuration to control the cutoff point? I doubt many users will need to change the default, but they might have already set `vite.build.assetsInlineLimit` explicitly for a different use-case.

**Reply by lilnasy** [reply]
@aFuzzyBear you've shown interest in this so please take a look

**Reply by matthewp** [reply]
My initial thoughts are I like it. Will provide more detailed feedback tomorrow.

**Reply by matthewp** [reply]
This has been accepted as is now a Stage 2 proposal. https://github.com/withastro/roadmap/issues/556

@lilnasy You can now take this comment and copy/paste it as a PR to this repo which constitutes the Stage 3 proposal.

From there we'll continue discussing, but presumably your PR will be merged as an experimental feature for some period of time before the Stage 3 is merged. Thanks again for championing this proposal.

**Comment by matthewp** [comment]
This has been accepted which closes out this issue. Continue discussion in https://github.com/withastro/roadmap/issues/556
